lisa/analysis/rta.py

```diff
@@ -15,187 +15,493 @@
+        :type phase: int
+
+        :returns: the requires task's phase end timestamp
+        """
+        return self._get_task_phase('end', task, phase)
+
+    @requires_events('rtapp_task')
+    def tasks_window(self, task):
+        """
+        Return the start end end time for the specified task.
+
+        :param task: Task names or PIDs to look for.
+        :type task: tupe(star, end)
+        """
+        task_df = self.df_rtapp_task(task)
+        start_time = task_df[task_df.event == "start"].index[0]
+        end_time = task_df[task_df.event == "end"].index[0]
+
+        return (start_time, end_time)
+
+    PhaseWindow = namedtuple("PhaseWindow", ['id', 'start', 'end'])
+
+    @df_rtapp_phases_start.used_events
+    def task_phase_window(self, task, phase):
+        """
+        Return the window of a requested task phase.
+
+        For the specified ``task`` it returns a tuple with the (start, end)
+        time of the requested ``phase``. A negative phase number can be used to
+        count phases backward from the last (-1) toward the first.
+
+        :param task: Task name or PID to look for
+        :type task: int or str
+
+        :param phase: The ID of the phase to consider
+        :type phase: int
+
+        :returns: tuple(start_s, end_s)
+        """
+        phase_start = self.df_rtapp_phase_start(task, phase)
+        phase_end = self.df_rtapp_phase_end(task, phase)
+
+        return self.PhaseWindow(phase, phase_start, phase_end)
+
+    ###########################################################################
+    # rtapp_phase events related methods
+    ###########################################################################
+
+    @requires_events('rtapp_event')
+    def df_rtapp_event(self, task=None):
+        df = self.trace.df_events('rtapp_event')
+        return self._task_filtered(df, task)
+
+    @df_rtapp_event.used_events
+    def df_rtapp_task_activations(self, task):
+        """
+        Dataframe of all activations of the specified task
+
+        :param task: the actual trace task name for an rt-app task to consider
+        :type task: str
+        """
+        df = self.df_rtapp_event(task)
+        return (df[(df["__comm"] == task) & (df.desc == "run")]
+                [["__cpu", "__pid"]])
+
+    ###########################################################################
+    # rtapp_stats events related methods
+    ###########################################################################
+
+    @memoized
+    @requires_events('rtapp_stats')
+    def _get_stats(self):
+        df = self.trace.df_events('rtapp_stats')
         # Add performance metrics column, performance is defined as:
         #             slack
         #   perf = -------------
         #          period - run
-        df['PerfIndex'] = df['Slack'] / (df['CPeriod'] - df['CRun'])
+        df.loc[:,'perf_index'] = df['slack'] / (df['c_period'] - df['c_run'])
```

Comment:
AFAIK `df.loc` will not create a copy of the dataframe on its own, which means this function will "corrupt" the dataframe by mutating it, so it should be copied first:
```
# Use shallow copy since we only want to add a new column without modifying the data/index of the existing ones.
df = df.copy(deep=False)
df.loc[...]
```

Otherwise we might end up with code relying on that column that seems to work because it happens to be called after that function, and will break the next day because it happens to be called before due to a totally unrelated change. 

note: If the callers of that function typically modify the index or data, we might as well make a deepcopy right here and let them play freely with the new independent df.

Reply:
So, either you get the DF using `rta.df_rtapp_stats()`, and you'll always get that column since it's memorized. Or, you get it via `trace.df_events()` and you'll never get the column.
The case where you use `rta.df_rtapp_stats()` and write some code that uses the new column and successively uses `rta.df_events()` honestly... looks like a very extreme corner case.

Are we really sure it makes sense to add memory overheads just for that?

Reply:
> Or, you get it via trace.df_events() and you'll never get the column.

Wouldn't `df.loc[:,'perf_index'] = ...` modify the dataframe returned by `df_events()` ? If it does modify it, then callers of `df_events()` may see that column.

> looks like a very extreme corner case.

Trace objects can be used in many parts of the code, making assumptions on the order in which things are done introduces hidden state that can only cause issues. If things don't have a good reason to be linked, they *must* not be linked. This kind of design leads to things like the now defunct `TestEnv` on which allocation of result directories was relying on a similar broken behavior. The code was written and documented to do something else, and what people wanted and were expecting to use was this "emerging" behavior.

> Are we really sure it makes sense to add memory overheads just for that?

I was suggesting using `deep=False`, which basically removes any memory-related concerns.

Reply:
>   Wouldn't df.loc[:,'perf_index'] = ... modify the dataframe returned by df_events() ? If it does modify it, then callers of df_events() may see that column.

Sure, but I don't expect those patterns if you are using the analysis API why should you mix it with a "raw" `df_events()`?

Reply:
> The code was written and documented to do something else, and what people wanted and were expecting to use was this "emerging" behavior.

I still see this as a very different case, if you access the data via `df_events()`  you are using a `raw` API and your code should not be considered stable by definition. There are no written rules we must maintain  a stable API for raw events.

Think about user-space injected events: if we go down the route you are tracing you commit yourself to keep backward/forward compatibility for all events accessed by `df_events()`.

My point is clear: if you use `df_events()` you are on your own. Especially when it comes to events supported by analysis APIs.

Reply:
My point was that two calls to `df_events()` with the same parameter on the same trace must return a dataframe with the same rows and same columns. Regardless of what happened in between. The data exposed by the `Trace` object are read-only, users do not have the right to modify it.